### ISSUE TITLE
Help fix uwsgi configuration issues with Django deployments

### ISSUE THREAD

**Issue by okhwaja** [bugsnag/bugsnag-python#157]
### Description
I ran into a difficult-to-debug issue when installing bugsnag and deploying my django application. Deploying using `uwsgi` (as described in [the official tutorial](https://docs.djangoproject.com/en/2.0/howto/deployment/wsgi/)), my application could not consistently send errors to bugsnag. Some would get through, but most seemed like they were hanging and then failing. The stack trace from the `uwsgi` suggested that the error was due to some ssl issues:

```
Traceback (most recent call last):
  File "/home/ubuntu/project/venv/lib/python3.5/site-packages/urllib3/connectionpool.py", line 600, in urlopen
    chunked=chunked)
  File "/home/ubuntu/project/venv/lib/python3.5/site-packages/urllib3/connectionpool.py", line 343, in _make_request
    self._validate_conn(conn)
  File "/home/ubuntu/project/venv/lib/python3.5/site-packages/urllib3/connectionpool.py", line 849, in _validate_conn
    conn.connect()
  File "/home/ubuntu/project/venv/lib/python3.5/site-packages/urllib3/connection.py", line 356, in connect
    ssl_context=context)
  File "/home/ubuntu/project/venv/lib/python3.5/site-packages/urllib3/util/ssl_.py", line 359, in ssl_wrap_socket
    return context.wrap_socket(sock, server_hostname=server_hostname)
  File "/usr/lib/python3.5/ssl.py", line 377, in wrap_socket
    _context=self)
  File "/usr/lib/python3.5/ssl.py", line 752, in __init__
    self.do_handshake()
  File "/usr/lib/python3.5/ssl.py", line 988, in do_handshake
    self._sslobj.do_handshake()
  File "/usr/lib/python3.5/ssl.py", line 633, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLEOFError: EOF occurred in violation of protocol (_ssl.c:645)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ubuntu/project/venv/lib/python3.5/site-packages/requests/adapters.py", line 445, in send
    timeout=timeout
  File "/home/ubuntu/project/venv/lib/python3.5/site-packages/urllib3/connectionpool.py", line 638, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/home/ubuntu/project/venv/lib/python3.5/site-packages/urllib3/util/retry.py", line 398, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='notify.bugsnag.com', port=443): Max retries exceeded with url: / (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:645)'),))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.5/threading.py", line 862, in run
    self._target(*self._args, **self._kwargs)
  File "/home/ubuntu/project/venv/lib/python3.5/site-packages/bugsnag/delivery.py", line 136, in request
    response = requests.post(uri, **req_options)
  File "/home/ubuntu/project/venv/lib/python3.5/site-packages/requests/api.py", line 112, in post
    return request('post', url, data=data, json=json, **kwargs)
  File "/home/ubuntu/project/venv/lib/python3.5/site-packages/requests/api.py", line 58, in request
    return session.request(method=method, url=url, **kwargs)
  File "/home/ubuntu/project/venv/lib/python3.5/site-packages/requests/sessions.py", line 512, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/ubuntu/project/venv/lib/python3.5/site-packages/requests/sessions.py", line 622, in send
    r = adapter.send(request, **kwargs)
  File "/home/ubuntu/project/venv/lib/python3.5/site-packages/requests/adapters.py", line 511, in send
    raise SSLError(e, request=request)
requests.exceptions.SSLError: HTTPSConnectionPool(host='notify.bugsnag.com', port=443): Max retries exceeded with url: / (Caused by SSLError(SSLEOFError(8, 'EOF occurred in violation of protocol (_ssl.c:645)'),))
```

Making this issue more confusing was that I was reliably sending messages using direct calls using the django shell: `import bugsnag; bugsnag.notify('...')` 

I eventually tried using Sentry after spending _a long time_ unsuccessfully debugging. Sentry didn't work either, but I very quickly noticed something interesting in my application logs:
```
/home/ubuntu/raven/venv/lib/python3.5/site-packages/raven/utils/compat.py:216: Warning: We detected the use of uwsgi with disabled threads.  This will cause issues with the transport you are trying to use.  Please enable threading for uwsgi.  (Enable the "enable-threads" flag).
  warn(Warning('We detected the use of uwsgi with disabled threads.  '
```

[source here](https://github.com/getsentry/raven-python/blob/master/raven/utils/compat.py#L216)

I also found a [side note in their documentation](https://docs.sentry.io/clients/python/advanced/#a-note-on-uwsgi) with some advice.

I adjusted my uwsgi.ini file like so:
```
...
 # skipping uid and guid
 harakiri=20
 max-requests=5000
+enable-threads=True
 vacuum=True
 daemonize=/home/ubuntu/out.log
 home=/home/ubuntu/project/venv
...
```

And Sentry worked. I went back to try Bugsnag and it worked too! 🎉 

**Describe the solution you'd like**

It'd be nice if Bugsnag made it easier to debug this issue going forward.
* Like Sentry's lib, an error/warning message indicating whether bugsnag will be able to create a thread needed to send exceptions
* Some section in the django/python documentation with a note about deployments with uwsgi to ensure threads are enabled.

I think it's probably worth addressing considering that the [django example deployment configuration](https://docs.djangoproject.com/en/2.0/howto/deployment/wsgi/uwsgi/) is not compatible with bugsnag as is and I think others like me will eventually stumble into this problem

**Additional context**
python 3.5.2
pip 10.0.1
Ubuntu 16.04

python packages
bugsnag=3.4.2
django=2.0.7
raven=6.9.0

**Comment by bengourley** [comment]
Hi @okhwaja. Thanks for taking the time to figure this out and for leaving such a detailed report.

I agree with both your points of action:

- updating the notifier to log a warning
- include the `enable-threads` setting in our uwsgi docs

I'll schedule this work.
Cheers!

**Comment by phillipsam** [comment]
An update for this has now been released in [v3.8.0](https://github.com/bugsnag/bugsnag-python/releases/tag/v3.8.0)
